src/Raven.Server/ServerWide/Context/TransactionOperationContext.cs

```diff
@@ -36,6 +42,79 @@ protected override RavenTransaction CreateWriteTransaction(TimeSpan? timeout = n
         {
             return new RavenTransaction(Environment.WriteTransaction(PersistentContext, Allocator, timeout));
         }
+
+        public void UpdatePrefixesOfReduceOutputDocumentsToDelete(ImmutableDictionary<string,string> prefixes)
```

Comment:
I have some doubts that `TransactionOperationContext` has those index specific method to deal with `IndexStateRecord`

Reply:
Yeah, but I was thinking about having something more generic way to play with the client state. Because now we have methods that update `IndexStateRecord` added to `TransactionOperationContext.cs` (e.g. `UpdatePrefixesOfReduceOutputDocumentsToDelete()`) and we call it via context - `context.UpdatePrefixesOfReduceOutputDocumentsToDelete`, but in other places you do it manually like:


- ConflictsStorage:
```
if (tx.LowLevelTransaction.TryGetClientState(out DocumentTransactionCache state) is false)
    state = new DocumentTransactionCache();
state.ConflictsCount++;
tx.LowLevelTransaction.UpdateClientState(state);
```

- RachisConsensus.cs:
```
public void UpdateStateIn(ClusterOperationContext ctx, RachisState state)
{
    if (ctx.Transaction.InnerTransaction.LowLevelTransaction.TryGetClientState(out ClusterStateRecord cur) is false)
        cur = ClusterStateRecord.Empty;
    ctx.Transaction.InnerTransaction.LowLevelTransaction.UpdateClientState(cur with
    {
        State = state,
        When = DateTime.UtcNow
    });
}
```


What if we go the following way:

- let's introduce an interface:


```
public interface IEnvironmentClientState
{
    static abstract IEnvironmentClientState CreateEmpty();
}
```

- then we can have generic methods to get (private) and update(public) the state:

```
public void UpdateClientState<TClientState>(UpdateClientStateDelegate<TClientState> update) where TClientState : IEnvironmentClientState
{
	var llt = Transaction.InnerTransaction.LowLevelTransaction;

	llt.UpdateClientState(update(GetClientState<TClientState>()));
}

private TClientState GetClientState<TClientState>() where TClientState : IEnvironmentClientState
{
	var llt = Transaction.InnerTransaction.LowLevelTransaction;

	if (llt.TryGetClientState(out TClientState r) is false)
		r = (TClientState)TClientState.CreateEmpty();

	return r;
}

public delegate TClientState UpdateClientStateDelegate<TClientState>(TClientState state) where TClientState : IEnvironmentClientState;
```

- the sample usage would be then:

```
context.UpdateClientState<IndexStateRecord>(state => state with { PrefixesOfReduceOutputDocumentsToDelete = _prefixesOfReduceOutputDocumentsToDelete.ToImmutable() });
```

Reply:
`UpdatePrefixesOfReduceOutputDocumentsToDelete` exists because I needed to call it from two places.
The others are in just one location. And the amount of code that this involves is pretty high.

Note that there is a _lot_ of code around that in `TransactionOperationContext` (which despite the name, is used for indexing only).

And modifying the state isn't really that simple. See for example: `ModifyReferencesState`.
That is often also performance sensitive, and we would like to avoid hitting that.

Reply:
Then maybe let's move this logic to the state itself? So we'd have:

```
context.UpdateClientState<IndexStateRecord>(state => state.UpdatePrefixesOfReduceOutputDocumentsToDelete(_prefixesOfReduceOutputDocumentsToDelete.ToImmutable()));
```

and the same we can do for `ModifyReferencesState()`.

My concern is that `TransactionOperationContext` is widely used across entire code base. It's not its resposibility to manage index specific state. The only thing we need from `TransactionOperationContext` is to access `LowLevelTransaction` and call `UpdateClientState()` internally.